Build failure on current Archlinux

I was trying to build from source just, and hit the error below. This is on current Archlinux. This bit looks relevant:

```
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.c++:17:
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.h:22:
    In file included from /ekam-provider/canonical/sandstorm/util.h:35:
    In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/set:61:
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_set.h:193:9: error: no matching member function for call to '_M_insert_range_unique'
            { _M_t._M_insert_range_unique(__first, __last); }
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~
```

But I haven't been able to sort out what's going on. Here's the full set of errors:

```
   /ekam-provider/canonical/sandstorm/supervisor.c++:791:19: warning: 'readdir_r' is deprecated [-Wdeprecated-declarations]
          int error = readdir_r(dir, &entry, &eptr);
                      ^
    /usr/include/dirent.h:186:28: note: 'readdir_r' has been explicitly marked deprecated here
         __nonnull ((1, 2, 3)) __attribute_deprecated__;
                               ^
    /usr/include/sys/cdefs.h:251:51: note: expanded from macro '__attribute_deprecated__'
    # define __attribute_deprecated__ __attribute__ ((__deprecated__))
                                                      ^
    /ekam-provider/canonical/sandstorm/supervisor.c++:1903:29: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
        return req.send().then([this, context](auto args) mutable -> void {
                                ^~~~~
    In file included from /ekam-provider/canonical/sandstorm/supervisor.c++:17:
    In file included from /ekam-provider/canonical/sandstorm/supervisor.h:20:
    In file included from /ekam-provider/canonical/sandstorm/abstract-main.h:20:
    In file included from /ekam-provider/c++header/kj/main.h:28:
    In file included from /ekam-provider/c++header/kj/array.h:28:
    /ekam-provider/c++header/kj/memory.h:405:60: warning: delete called on non-final 'sandstorm::SupervisorMain::SupervisorImpl::LogWatcher' that has virtual functions but non-virtual destructor [-Wdelete-non-
      virtual-dtor]
      virtual void disposeImpl(void* pointer) const override { delete reinterpret_cast<T*>(pointer); }
                                                               ^
    /ekam-provider/c++header/kj/memory.h:411:51: note: in instantiation of member function 'kj::_::HeapDisposer<sandstorm::SupervisorMain::SupervisorImpl::LogWatcher>::disposeImpl' requested here
    const HeapDisposer<T> HeapDisposer<T>::instance = HeapDisposer<T>();
                                                      ^
    /ekam-provider/c++header/kj/memory.h:422:72: note: in instantiation of static data member 'kj::_::HeapDisposer<sandstorm::SupervisorMain::SupervisorImpl::LogWatcher>::instance' requested here
      return Own<T>(new T(kj::fwd<Params>(params)...), _::HeapDisposer<T>::instance);
                                                                           ^
    /ekam-provider/canonical/sandstorm/supervisor.c++:2099:24: note: in instantiation of function template specialization 'kj::heap<sandstorm::SupervisorMain::SupervisorImpl::LogWatcher, kj::UnixEventPort &, char
      const (&)[4], kj::AutoCloseFd, sandstorm::ByteStream::Client>' requested here
        auto watcher = kj::heap<LogWatcher>(eventPort, "log", kj::mv(logFile), params.getStream());
    ...(log truncated; use -l to increase log limit)...
✔ test: capnp/fuzz-test
✘ compile: sandstorm/web-session-bridge.c++
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:143:35: warning: lambda capture 'method' is not used [-Wunused-lambda-capture]
          auto doNonStreaming = [this,method,path,&headers,&requestBody,&response]() {
                                     ~^~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:173:24: warning: lambda capture 'headers' is not used [-Wunused-lambda-capture]
              .then([this,&headers,&requestBody,&response,KJ_MVCAP(streamer)]
                         ~~^~~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:177:11: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
          }, [this,KJ_MVCAP(doNonStreaming)](kj::Exception&& e) -> kj::Promise<void> {
              ^~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:192:35: warning: lambda capture 'method' is not used [-Wunused-lambda-capture]
          auto doNonStreaming = [this,method,path,&headers,&requestBody,&response]() {
                                     ~^~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:222:24: warning: lambda capture 'headers' is not used [-Wunused-lambda-capture]
              .then([this,&headers,&requestBody,&response,KJ_MVCAP(streamer)]
                         ~~^~~~~~~
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:226:11: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
          }, [this,KJ_MVCAP(doNonStreaming)](kj::Exception&& e) -> kj::Promise<void> {
              ^~~~~
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.c++:17:
    In file included from /ekam-provider/canonical/sandstorm/web-session-bridge.h:22:
    In file included from /ekam-provider/canonical/sandstorm/util.h:35:
    In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/set:61:
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_set.h:193:9: error: no matching member function for call to '_M_insert_range_unique'
            { _M_t._M_insert_range_unique(__first, __last); }
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~
    /ekam-provider/canonical/sandstorm/util.h:244:30: note: in instantiation of function template specialization 'std::set<kj::StringPtr, std::less<kj::StringPtr>, std::allocator<kj::StringPtr> >::set<capnp::_::
      IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader> >' requested here
      HeaderWhitelist(T&& list): patterns(list.begin(), list.end()) {}
                                 ^
    /ekam-provider/canonical/sandstorm/web-session-bridge.c++:98:7: note: in instantiation of function template specialization 'sandstorm::HeaderWhitelist::HeaderWhitelist<capnp::List<capnp::Text, capnp::Kind::
    ...(log truncated; use -l to increase log limit)...
✘ compile: sandstorm/sandstorm-http-bridge.c++
    /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:1520:10: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
            [this, KJ_MVCAP(httpRequest), KJ_MVCAP(clientStream), responseStream, context]
             ^~~~~
    /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:2468:37: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
        return address->connect().then([this, loggedSlowStartupMessage](auto x) -> void {
                                        ^~~~~
    In file included from /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:59:
    In file included from /ekam-provider/canonical/sandstorm/util.h:35:
    In file included from /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/set:61:
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_set.h:193:9: error: no matching member function for call to '_M_insert_range_unique'
            { _M_t._M_insert_range_unique(__first, __last); }
              ~~~~~^~~~~~~~~~~~~~~~~~~~~~
    /ekam-provider/canonical/sandstorm/util.h:244:30: note: in instantiation of function template specialization 'std::set<kj::StringPtr, std::less<kj::StringPtr>, std::allocator<kj::StringPtr> >::set<capnp::_::
      IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader> >' requested here
      HeaderWhitelist(T&& list): patterns(list.begin(), list.end()) {}
                                 ^
    /ekam-provider/canonical/sandstorm/sandstorm-http-bridge.c++:154:23: note: in instantiation of function template specialization 'sandstorm::HeaderWhitelist::HeaderWhitelist<capnp::List<capnp::Text, capnp::
      Kind::BLOB>::Reader>' requested here
    const HeaderWhitelist REQUEST_HEADER_WHITELIST(*WebSession::Context::HEADER_WHITELIST);
                          ^
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_tree.h:1109:2: note: candidate template ignored: substitution failure [with _InputIterator = capnp::_::IndexingIterator<const 
      capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader>]: no type named 'value_type' in 'std::iterator_traits<capnp::_::IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::
      Reader, capnp::Text::Reader> >'
            _M_insert_range_unique(_InputIterator __first, _InputIterator __last)
            ^
    /usr/lib64/gcc/x86_64-pc-linux-gnu/9.2.0/../../../../include/c++/9.2.0/bits/stl_tree.h:1118:2: note: candidate template ignored: substitution failure [with _InputIterator = capnp::_::IndexingIterator<const 
      capnp::List<capnp::Text, capnp::Kind::BLOB>::Reader, capnp::Text::Reader>]: no type named 'value_type' in 'std::iterator_traits<capnp::_::IndexingIterator<const capnp::List<capnp::Text, capnp::Kind::BLOB>::
      Reader, capnp::Text::Reader> >'
            _M_insert_range_unique(_InputIterator __first, _InputIterator __last)
            ^
    ...(log truncated; use -l to increase log limit)...
✘ link: sandstorm/run-bundle.o
    /usr/bin/ld: tmp/kj/async-io-unix.o: in function `operator()':
    /ekam-provider/canonical/kj/async-io-unix.c++:989: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `Tables':
    /ekam-provider/canonical/sandstorm/gateway.c++:92: undefined reference to `sandstorm::WebSessionBridge::Tables::Tables(kj::HttpHeaderTable::Builder&)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `sandstorm::GatewayService::request(kj::HttpMethod, kj::StringPtr, kj::HttpHeaders const&, kj::AsyncInputStream&, kj::HttpService::Response&)':
    /ekam-provider/canonical/sandstorm/gateway.c++:222: undefined reference to `sandstorm::WebSessionBridge::addStandardApiOptions(sandstorm::WebSessionBridge::Tables const&, kj::HttpHeaders const&, kj::
      HttpHeaders&)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `sandstorm::GatewayService::getStaticPublished(kj::StringPtr, kj::StringPtr, kj::HttpHeaders const&, kj::HttpService::Response&, unsigned int)':
    /ekam-provider/canonical/sandstorm/gateway.c++:759: undefined reference to `sandstorm::WebSessionBridge::makeHttpResponseStream(unsigned int, kj::StringPtr, kj::HttpHeaders&&, kj::HttpService::Response&)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `operator()':
    /ekam-provider/canonical/sandstorm/gateway.c++:270: undefined reference to `sandstorm::WebSessionBridge::addStandardApiOptions(sandstorm::WebSessionBridge::Tables const&, kj::HttpHeaders const&, kj::
      HttpHeaders&)'
    /usr/bin/ld: /ekam-provider/canonical/sandstorm/gateway.c++:489: undefined reference to `sandstorm::WebSessionBridge::restrictParentFrame(kj::StringPtr, kj::StringPtr)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `kj::Own<sandstorm::WebSessionBridge> kj::refcounted<sandstorm::WebSessionBridge, kj::Timer&, sandstorm::WebSession::Client, sandstorm::Handle::Client, 
      sandstorm::WebSessionBridge::Tables&, sandstorm::WebSessionBridge::Options&>(kj::Timer&, sandstorm::WebSession::Client&&, sandstorm::Handle::Client&&, sandstorm::WebSessionBridge::Tables&, sandstorm::
      WebSessionBridge::Options&)':
    /ekam-provider/c++header/kj/refcount.h:97: undefined reference to `sandstorm::WebSessionBridge::WebSessionBridge(kj::Timer&, sandstorm::WebSession::Client, kj::Maybe<sandstorm::Handle::Client>, sandstorm::
      WebSessionBridge::Tables const&, sandstorm::WebSessionBridge::Options)'
    /usr/bin/ld: tmp/sandstorm/gateway.o: in function `kj::Own<sandstorm::WebSessionBridge> kj::refcounted<sandstorm::WebSessionBridge, kj::Timer&, sandstorm::WebSession::Client, decltype(nullptr), sandstorm::
      WebSessionBridge::Tables&, sandstorm::WebSessionBridge::Options&>(kj::Timer&, sandstorm::WebSession::Client&&, decltype(nullptr)&&, sandstorm::WebSessionBridge::Tables&, sandstorm::WebSessionBridge::Options
      &)':
    /ekam-provider/c++header/kj/refcount.h:97: undefined reference to `sandstorm::WebSessionBridge::WebSessionBridge(kj::Timer&, sandstorm::WebSession::Client, kj::Maybe<sandstorm::Handle::Client>, sandstorm::
      WebSessionBridge::Tables const&, sandstorm::WebSessionBridge::Options)'
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
✘ install: sandstorm/sandstorm.ekam-manifest
    sandstorm/run-bundle: not found
vim 
```

--- a/Makefile
+++ b/Makefile
@@ -31,12 +31,16 @@ EKAM=ekam
 #   needs to include all the Cap'n Proto code. Do we double-compile or do we
 #   just accept it? Perhaps it's for the best since we probably should build
 #   position-independent executables for security reasons?
+# NOTE: Emperically -DKJ_STD_COMPAT appears to be necessary on recent ditros
+#   (as of Jan 2020), notably Debian Sid, Fedora 30 and Archlinux. We're not
+#   entirely sure what changed, but this seems like a backwards incompatibility
+#   in libstdc++. See also issue #3171.
 METEOR_DEV_BUNDLE=$(shell ./find-meteor-dev-bundle.sh)
 NODEJS=$(METEOR_DEV_BUNDLE)/bin/node
 NODE_HEADERS=$(METEOR_DEV_BUNDLE)/include/node
 WARNINGS=-Wall -Wextra -Wglobal-constructors -Wno-sign-compare -Wno-unused-parameter
-CXXFLAGS2=-std=c++1z $(WARNINGS) $(CXXFLAGS) -DSANDSTORM_BUILD=$(BUILD) -DKJ_HAS_OPENSSL -DKJ_HAS_ZLIB -DKJ_HAS_LIBDL -pthread -fPIC -I$(NODE_HEADERS)
-CFLAGS2=$(CFLAGS) -pthread -fPIC
+CXXFLAGS2=-std=c++1z $(WARNINGS) $(CXXFLAGS) -DSANDSTORM_BUILD=$(BUILD) -DKJ_HAS_OPENSSL -DKJ_HAS_ZLIB -DKJ_HAS_LIBDL -pthread -fPIC -I$(NODE_HEADERS) -DKJ_STD_COMPAT
+CFLAGS2=$(CFLAGS) -pthread -fPIC -DKJ_STD_COMPAT
 LIBS2=$(LIBS) deps/libsodium/build/src/libsodium/.libs/libsodium.a deps/boringssl/build/ssl/libssl.a deps/boringssl/build/crypto/libcrypto.a -lz -ldl -pthread
 
 define color

